src/Microsoft.DotNet.ServiceFabric.ServiceHost/ServiceHostEventListener.cs

```diff
@@ -0,0 +1,147 @@
+{
+    private static ServiceHostEventListener _instance;
+    private static IImmutableList<string> _eventSourceNamePrefixes;
+    private static ILoggerFactory _loggerFactory;
+    private static ConcurrentDictionary<string, ILogger> _loggers = new();
+
+    public static ServiceHostEventListener ListenToEventSources(ILoggerFactory loggerFactory, params string[] namePrefixes)
+    {
+        if (_instance is not null)
+        {
+            throw new InvalidOperationException("Cannot call ServiceHostEventListener.ListenToEventSources more than once.");
+        }
+        // These fields must be static because the "OnEventSourceCreated" method gets called inside the base constructor, so instance fields aren't initialized when it runs
+        _eventSourceNamePrefixes = namePrefixes.ToImmutableList();
+        _loggerFactory = loggerFactory;
+        return _instance = new ServiceHostEventListener();
+    }
+
+
+    protected override void OnEventSourceCreated(EventSource eventSource)
+    {
+        if (_eventSourceNamePrefixes.Any(prefix => eventSource.Name.StartsWith(prefix)))
+        {
+            EnableEvents(eventSource, EventLevel.Warning);
+            _loggers.TryAdd(eventSource.Name, _loggerFactory.CreateLogger(eventSource.Name));
+        }
+        base.OnEventSourceCreated(eventSource);
+    }
+
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        var level = eventData.Level switch
+        {
+            EventLevel.LogAlways => LogLevel.Critical,
+            EventLevel.Critical => LogLevel.Critical,
+            EventLevel.Error => LogLevel.Error,
+            EventLevel.Warning => LogLevel.Warning,
+            EventLevel.Informational => LogLevel.Information,
+            EventLevel.Verbose => LogLevel.Trace,
+            _ => LogLevel.None,
+        };
+
+        if (!_loggers.TryGetValue(eventData.EventSource.Name, out var logger))
+        {
+            return;
+        }
+
+        if (eventData.PayloadNames is not null && eventData.Payload is not null)
+        {
+            logger.Log(level, 
+                0,
+                new EventListenerLogValues(eventData.EventName, eventData.Message, eventData.PayloadNames, eventData.Payload), 
+                null,
+                (v, _) => v.ToString());
+        }
+        else if (eventData.Payload is not null)
+        {
+            var sb = new StringBuilder(4 + eventData.EventName.Length + eventData.Message.Length + eventData.Payload.Count * 8);
+            sb.Append(eventData.EventName);
+            sb.Append(": ");
+            sb.AppendFormat(eventData.Message, eventData.Payload.ToArray());
+            sb.Append("; ");
+            for (int i = 0; i < eventData.Payload.Count; i++)
+            {
+                if (i != 0)
+                {
+                    sb.Append(", ");
+                }
+                sb.Append(eventData.Payload[i]);
+            }
+
+            logger.Log(level, sb.ToString());
```

Comment:
Doesn't the logger already handle the format strings? Why are we manually doing the formatting ourselves?

Reply:
The EventSource format strings aren't "logger" format strings. They are "This {2} thing {1} did {0}". The parameters would all get switched around.

Reply:
Oooo, it's the old kind. Gotcha. :-)